### ISSUE TITLE
Bug: [no-misused-promises] bad report loc for methods in object literals

### ISSUE THREAD

**Issue by kirkwaiblinger** [typescript-eslint/typescript-eslint#10212]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.6.2&showAST=ts&fileType=.ts&code=C4TwDgpgBAEghgZwGoHsCWATAshYALFDKAXigG8AoKKAMwAoBKALigDd0MKBfCigYxQA7BMCgoW8ZBxz5CJclSiIQgvrUbkoPLkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oFtLlZkiACa1i0Dr0GoMkAO4BDaE0jgwAXxBqgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
type HasVoidMethod = {
  f(): void
}

const o: HasVoidMethod = {
  async f() { }
}
```

### ESLint Config

```javascript
module.exports = {
  "rules": {
    "@typescript-eslint/no-misused-promises": "warn"
  }
}
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

```ts
const o: HasVoidMethod = {
  async f() { }
//~~~~~~~
}
```

### Actual Result

```ts
const o: HasVoidMethod = {
  async f() { }
//       ~~~~~~
}
```

### Additional Info

For methods, should probably underline an explicitly provided return type if present. _Otherwise_, use the getFunctionHeadLoc utility.

**Comment by Gyumong** [comment]
If no one is currently assigned, I would like to work on resolving this issue.

**Comment by kirkwaiblinger** [comment]
Go for it! We invite contributions for things marked "accepting prs" 🙂

https://typescript-eslint.io/contributing/pull-requests
